### PR TITLE
ES6hasInstance

### PR DESCRIPTION
**Pull request by suwc**
Recuperate perf loss from existing ES6hasInstance implementation.
Perf-neutral with vs. without -es6hasinstance switch.
Add inline-cache-invalidation for user-defined Symbol.hasInstance property in function constructor.
Revamp unit tests for ES6hasInstance.

Related ECMA262 spec sections:
6.1.5.1 Well-Known Symbols
7.3.19 OrdinaryHasInstance (C, O)
12.10.4 Runtime Semantics: InstanceofOperator(O, C)
19.2.3.6 Function.prototype [ @@hasInstance ] ( V )

### CODE REVIEW

**Comment by suwc** [comment]
@Microsoft/chakra-es could you take a look pls?

**Comment by abchatra** [comment]
>         if (scriptContext->GetConfig()->IsES6HasInstanceEnabled())

Add Symbol.hasInstance as part of experimental flags. 

---
Refers to: lib/Runtime/Library/JavascriptLibrary.cpp:2544 in 79f7012. [](commit_id = 79f7012c085fc9a21441cd6da9d9367d4e29d8d3, deletion_comment = False)

File: test/es6/es6HasInstance.js [hunk 1]
```diff
@@ -3,86 +3,396 @@
...
+
         }
-    });
+    },
+];
 
-    var x = proxy.bind();
-    print((x) instanceof proxy);
-}
+testRunner.runTests(tests, { verbose: WScript.Arguments[0] != "summary" });
```
Comment (abchatra) [inline test/es6/es6HasInstance.js]: Please run test262 to see if there any additional failures in Symbol.hasInstance

**Comment by abchatra** [comment]
:shipit:

File: test/es6/es6HasInstance.js (hunk 1 above)
Comment (suwc) [inline test/es6/es6HasInstance.js]: Confirmed no test262 or perf regressions.

---
In reply to: [65278274](https://github.com/Microsoft/ChakraCore/pull/1063#discussion_r65278274) [](ancestors = 65278274)

**Comment by suwc** [comment]
>         if (scriptContext->GetConfig()->IsES6HasInstanceEnabled())

Done.

---
In reply to: [222847263](https://github.com/Microsoft/ChakraCore/pull/1063#issuecomment-222847263) [](ancestors = 222847263)

---
Refers to: lib/Runtime/Library/JavascriptLibrary.cpp:2544 in 79f7012. [](commit_id = 79f7012c085fc9a21441cd6da9d9367d4e29d8d3, deletion_comment = False)

**Pull request merged**